### PR TITLE
fix(core): validate session variable assignments

### PR DESCRIPTION
**Pull request by denyspakizh-tob**
## Pull Request Checklist

- [x] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? 
- [ ] Did you update the typescript typings accordingly (if applicable)? 
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?

## Description of Changes

`Sequelize#setSessionVariables` interpolated variable names and values directly into a single
`SET` statement. An attacker-controlled value could add assignments or subqueries without
requiring `multipleStatements`.

Values now pass through Sequelize's normal MySQL/MariaDB escaping path. This closes the injection
under the default SQL modes. The shared escaper's pre-existing `NO_BACKSLASH_ESCAPES` limitation is
outside this PR.

Names are validated against the supported 1-64 character unquoted user-variable grammar,
including `.` and U+0080-U+FFFF. SQL expressions are rejected because this API accepts values,
not raw SQL fragments.

Added integration coverage for normal values, valid and invalid names, and expression rejection,
plus an always-on unit assertion that the generated SQL keeps an injection value inert.

Tests run:

- `DIALECT=mariadb yarn workspace @sequelize/core mocha "test/unit/sequelize.test.ts" "test/integration/sequelize/set-session-variables.test.ts"`
- `DIALECT=mysql yarn workspace @sequelize/core mocha "test/unit/sequelize.test.ts"`
- `tsc -b packages/core/test/tsconfig.json`
- `eslint` and `prettier --check` on changed files

## List of Breaking Changes

Quoted user-variable names, such as `` `my var` ``, `'x'`, or `"x"`, are now rejected. Use an
unquoted name matching the supported grammar.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved MySQL/MariaDB `SET` session-variable handling with stricter validation of variable names.
  * Rejects invalid session variable names and prevents SQL-expression-like values from being used as values.
  * Produces safer `SET` statements for string values, correctly escaping quotes and backslashes (including `null` and scalar values).
* **Tests**
  * Expanded unit and integration coverage for single/multiple variables, dollar-prefixed keys, escaping round-trips, and rejection cases.

### CODE REVIEW

File: packages/core/src/sequelize.js [hunk 1]
```diff
@@ -63,6 +63,8 @@ import { withSqliteForeignKeysOff } from './utils/sql.js';
 import { useInflection } from './utils/string';
 import { Validator } from './utils/validator-extras';
 
+const MYSQL_MARIADB_USER_VARIABLE_NAME = /^[a-zA-Z0-9_$.]{1,64}$/;
```
Comment (WikiRik) [inline packages/core/src/sequelize.js]: This rejects four name shapes that MySQL accepts and that work on `main` today — I verified all four on 8.0.19:

| name | server | this PR |
|---|---|---|
| `` @`my var` `` | works | rejected |
| `@'x'` / `@"x"` | works | rejected |
| `@café` | works | rejected |

So the PR description's "no breaking changes" isn't quite right. It's a *small* break and arguably an acceptable one, but it should be stated.

If you want to keep the allowlist and just widen it to match MySQL 8's actual grammar (identifier characters plus `.`, including U+0080–U+FFFF):

```suggestion
const MYSQL_MARIADB_USER_VARIABLE_NAME = /^[a-zA-Z0-9_$.\u0080-\uFFFF]{1,64}$/;
```

Please write the high range with the explicit `\u0080` escape rather than a literal character — a raw U+0080 is invisible in most editors and renders the class as though it were the range `.`–`\uFFFF`, which is alarming and wrong. (I know because it briefly fooled a reviewer here.)

The alternative — `quoteIdentifier(name)` instead of an allowlist — removes the "is the allowlist complete?" question entirely, but it isn't free either: quoting changes *which* variable the value lands in, so an app reading `@my var` would silently get `NULL`. Trading a loud `TypeError` for a silent `NULL` may well be worse. Your call; I'd lean toward the widened allowlist.

File: packages/core/src/sequelize.js [hunk 2]
```diff
@@ -431,10 +433,21 @@ Use Sequelize#query if you wish to use replacements.`);
...
+        );
+      }
+
+      if (value instanceof BaseSqlExpression || (isPlainObject(value) && Op.col in value)) {
+        throw new TypeError(
+          'sequelize.setSessionVariables does not accept SQL expressions as variable values',
+        );
+      }
+
+      return `@${key} := ${this.escape(value)}`;
```
Comment (WikiRik) [inline packages/core/src/sequelize.js:449]: Routing values through `escape()` is the right fix and closes the injection under default `sql_mode`.

One residual worth knowing about, **not something to fix here**: under `sql_mode=NO_BACKSLASH_ESCAPES`, `escapeMysqlMariaDbString` still emits `\'`, which the server treats as two literal characters — so the value remains injectable in that configuration. I verified this on both MySQL 8.0.19 and MariaDB 10.11.

That's a pre-existing, repo-wide issue with the escaper itself and it's being tracked separately. The only ask here is that the PR description not claim this fully closes the injection in all configurations — it closes it in the default one.

**Review by WikiRik — COMMENTED** [review]
This is a real fix for a real vulnerability, and the validation holds — I couldn't find a bypass.

I confirmed exploitation on `main` against MySQL 8.0.19: an attacker-controlled **value** such as `", @is_admin := 1, @leak := (SELECT password FROM users LIMIT 1), @z := "` escalates a privilege variable and exfiltrates a password hash. Worth emphasising that this needs **no** `multipleStatements` — the injection stays inside the single `SET`, so the usual mysql2 mitigation doesn't apply. The values are the serious half; the name allowlist is defence-in-depth.

I probed the new regex hard and it holds up: `flags === ''` so `$` asserts end-of-input (JS is not Python/Ruby here — `/^a$/.test('a\n')` is `false`), `\n` isn't in the class anyway, and ReDoS is structurally impossible with a single bounded class. 36 adversarial names — CR/CRLF/U+2028/U+0085/NUL/backtick/`--`/`/* */`/emoji — were all rejected.

Two things to flag, both inline:

1. The allowlist is a **small breaking change** the description says it isn't: `` @`my var` ``, `@'x'`, `@"x"` and `@café` are all accepted by MySQL 8.0.19 and work on `main` today.
2. A residual: `escape()` is still bypassable under `sql_mode=NO_BACKSLASH_ESCAPES`. That's pre-existing and **not this PR's job** — but it means this PR shouldn't be described as fully closing the injection on its own.

Also: the regression tests are integration-only, so they're skipped outside the mysql/mariadb jobs, and none asserts the generated SQL. A unit-level assertion that an injection payload produces inert SQL would be the highest-value test here and would run in the always-on job.

One process note for maintainers rather than for you: given this is reachable from ordinary request data, it should ship with an advisory and a release note rather than a plain `fix(core)` entry.

---
*Reviewed by Opus 5 with Claude Code, supervised by @WikiRik.*

**Comment by denyspakizh-tob** [comment]
`99eeb9d34` adds generated-SQL coverage, broadens valid names, and documents the SQL-mode scope and quoted-name break. The TypeScript extraction was skipped, please let me know if it must be taken care of here

**Comment by denyspakizh-tob** [comment]
Fixed in `a9c21c103`; the unit test now asserts apostrophe and backslash escaping. MySQL/MariaDB tests, types, lint, and format pass.

File: packages/core/src/sequelize.js [hunk 3]
```diff
@@ -63,6 +63,8 @@ import { withSqliteForeignKeysOff } from './utils/sql.js';
 import { useInflection } from './utils/string';
 import { Validator } from './utils/validator-extras';
 
+const MYSQL_MARIADB_USER_VARIABLE_NAME = /^[a-zA-Z0-9_$.]{1,64}$/;
+
```
[the remark this answers is not in the document]
Comment (WikiRik) [reply packages/core/src/sequelize.js]: Usually I would agree, but in this case the benefit is greater to merge it as is and we move it to TypeScript later

**Review by WikiRik — APPROVED** [review]
